craft_application/services/config.py

```diff
@@ -0,0 +1,197 @@
+#  This file is part of craft-application.
+#
+#  Copyright 2024 Canonical Ltd.
+#
+#  This program is free software: you can redistribute it and/or modify it
+#  under the terms of the GNU Lesser General Public License version 3, as
+#  published by the Free Software Foundation.
+#
+#  This program is distributed in the hope that it will be useful, but WITHOUT
+#  ANY WARRANTY; without even the implied warranties of MERCHANTABILITY,
+#  SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR PURPOSE.
+#  See the GNU Lesser General Public License for more details.
+#
+#  You should have received a copy of the GNU Lesser General Public License
+#  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""Configuration service."""
+from __future__ import annotations
+
+import abc
+import contextlib
+import enum
+import os
+from collections.abc import Iterable
+from typing import TYPE_CHECKING, Any, TypeVar, cast, final
+
+import pydantic
+import pydantic_core
+import snaphelpers
+from craft_cli import emit
+from typing_extensions import override
+
+from craft_application import _config, application, util
+from craft_application.services import base
+
+if TYPE_CHECKING:
+    from craft_application.services.service_factory import ServiceFactory
+
+
+T = TypeVar("T")
+
+
+class ConfigHandler(abc.ABC):
+    """An abstract class for configuration handlers."""
+
+    def __init__(self, app: application.AppMetadata) -> None:
+        self._app = app
+
+    @abc.abstractmethod
+    def get_raw(self, item: str) -> Any:  # noqa: ANN401
+        """Get the string value for a configuration item.
+
+        :param item: the name of the configuration item.
+        :returns: The raw value of the item.
+        :raises: KeyError if the item cannot be found.
+        """
+
+
+@final
+class AppEnvironmentHandler(ConfigHandler):
+    """Configuration handler to get values from app-specific environment variables."""
+
+    def __init__(self, app: application.AppMetadata) -> None:
+        super().__init__(app)
+        self._environ_prefix = f"{app.name.upper()}"
+
+    @override
+    def get_raw(self, item: str) -> str:
+        return os.environ[f"{self._environ_prefix}_{item.upper()}"]
+
+
+@final
+class CraftEnvironmentHandler(ConfigHandler):
+    """Configuration handler to get values from CRAFT environment variables."""
+
+    def __init__(self, app: application.AppMetadata) -> None:
+        super().__init__(app)
+        self._fields = _config.ConfigModel.model_fields
+
+    @override
+    def get_raw(self, item: str) -> str:
+        # Ensure that CRAFT_* env vars can only be used for configuration items
```

Comment:
Does it also affect the `CRAFT_STORE*` env vars ?

Reply:
Yes. However, one could make a `CraftStoreEnvironmentHandler` that gets `CRAFT_STORE_*` variables. However, that would be a bit weird since craft-store has no awareness of craft-application.